opencv/opencv
# [GSoC] OpenCV.js: Accelerate OpenCV.js DNN via WebNN

### Overview

Proposal: OpenCV.js: Accelerate OpenCV.js DNN via WebNN
Mentor: Ningxin Hu @huningxin
Student: Hanxi Guo @MarkGHX

<cut/>

### This pull request changes

- Modify the CMAKE files to fit WebNN.
- Create the WebNN backend prototype for OpenCV.js dnn module in C++.
- Create the WebNN ReLULayer and pass the test case.
- Create the WebNN implementation of selected compute-intensive operations.
  - BatchNormLayer
  - ConvolutionLayer (group = 1)
  - FullyConnectedLayer
  - ReLULayer
  - ReLU6Layer
  - PoolingLayer
  - ReshapeLayer
  - SoftmaxLayer
  - PermuteLayer
  - ConstantLayer
  - ConcatLayer
- Compile the OpenCV.js dnn WebNN backend implementation into WebAssembly.
- Create the image classification sample of OpenCV.js dnn module using WebNN Backend, collect and analyze the performance numbers of different implementations including WebAssembly, WebNN (both WebNN-polyfill and electron) and native.

### Test
#### My test environments
```
OS: Ubuntu Linux 18.04.5 LTS
Emscripten: 2.0.15
Browser: Chrome, Version 90.0.4430.72 (Official Build) (64-bit)
Hardware: Intel® Core™ i9-7920X CPU @ 2.90GHz and GeForce GTX 1060 3GB
```
#### Preparations
1. Download this PR to `/GSoC2021` (for example).
2. Download [emsdk](https://github.com/emscripten-core/emsdk) to `/GSoC2021` and install version `2.0.15`.
3. Replace `/GSoC2021/emsdk/upstream/emscripten` with [emscripten-webnn](https://github.com/huningxin/emscripten) (branch `webnn_2.0.15`).
#### To run OpenCV native DNN module with WebNN backend
1. Refer to [WebNN's build instructions](https://github.com/webmachinelearning/webnn-native) to complete the build of WebNN-native with OpenVINO.
2. Download [opencv_extra](https://github.com/opencv/opencv_extra).
3. Set the environment variables:
```shell
$ source /opt/intel/<openvino_dir>/bin/setupvars.sh
$ export WEBNN_NATIVE_DIR=<webnn_native_out_dir>
$ export LD_LIBRARY_PATH=<opencv_dir>/lib:<webnn_native_out_dir>:${LD_LIBRARY_PATH}
$ export OPENCV_TEST_DATA_PATH=<opencv_extra_dir>/testdata/
```
4. Build OpenCV native with WebNN backend:
```shell
$ mkdir build
$ cd build
$ cmake -DCMAKE_BUILD_TYPE=Release -DWITH_WEBNN=ON -DBUILD_EXAMPLES=ON -DBUILD_TEST=ON -DCMAKE_INSTALL_PREFIX=/usr/local ..
$ make
```
5. Copy model files and figures to `./opencv/build/bin`
6. Run the OpenCV native DNN module with WebNN backend:
```shell
$ ./bin/example_dnn_classification --model=./bin/googlenet-v1.caffemodel --config=./bin/googlenet-v1.prototxt --width=224 --height=224 --classes=./bin/classification_classes_ILSVRC2012.txt --input=./bin/space_shuttle.jpg --mean="104 117 123" --rgb=false --backend=6
```
7.Expected result:
![image](https://user-images.githubusercontent.com/35370837/131092355-4b453b2f-9483-4d89-ba61-e041c4881355.png)
#### To run OpenCV.js DNN module with WebNN backend using WebNN-polyfill
1. Build WebNN-polyfill and OpenCV.js:
```shell
$ cd GSoC2021
$ git clone https://github.com/MarkGHX/webnn-polyfill
$ cd webnn-polyfill/
$ npm install
$ npm run build
$ cd ..
$ cd opencv
$ mkdir build_js
$ emcmake python ./platforms/js/build_js.py build_js --build_wasm --build_loader --build_doc --build_test --webnn
$ cd ..
$ cd webnn-polyfill/
$ cp dist/webnn-polyfill.js ~/GSoC2021/opencv/build_js/doc/doxygen/html/
$ cd ..
```
2. Set up http-server:
```shell
$ cd opencv/build_js/doc/doxygen/html/
$ http-server
```
3. Open Chrome and go to `http://127.0.0.1:8080/js_image_classification_webnn_polyfill.html`. Then you can test OpenCV.js GoogleNet with WebNN backend in image classification task.
4. Expected results:
![polyfill](https://user-images.githubusercontent.com/35370837/133200951-6e0b35d1-8e08-4b65-918c-93fa1a5434b9.png)

#### To run OpenCV.js DNN module with WebNN backend using Electron
1. Refer to [WebNN's build instructions](https://github.com/webmachinelearning/webnn-native) to complete the build of WebNN-native with OpenVINO.
2. Download [opencv_extra](https://github.com/opencv/opencv_extra).
3. Set the environment variables:
```shell
$ source /opt/intel/<openvino_dir>/bin/setupvars.sh
$ export WEBNN_NATIVE_DIR=<webnn_native_out_dir>
$ export LD_LIBRARY_PATH=./lib:<webnn_native_out_dir>:${LD_LIBRARY_PATH}
$ export OPENCV_TEST_DATA_PATH=<opencv_extra_dir>/testdata/
```
4. Build electron in OpenCV.js DNN module folder:
```shell
$ cd opencv/build_js/doc/doxygen/html/
$ npm install
$ npm run start
```
5. Expected result:
![electron](https://user-images.githubusercontent.com/35370837/133201022-ecd0309c-79f8-4b60-801c-315ec5696fcf.png)

### Performance results
#### Inference time in one round
| Model   |     OpenCV.js wasm      |  OpenCV.js wasm+simd+threads | OpenCV native default | OpenCV OpenVINO | OpenCV WebNN | OpenCV.js WebNN-polyfill | OpenCV.js WebNN-Electron |
|:----------:|:-------------:|:------:|:----------:|:---------:|:---------:|:--------:|:------------:|
| GoogleNet | 825.07ms | 51.55ms | 29.32ms | 10.35ms | 24.8ms | 69.15ms | 24.90ms |
| SqueezeNet | 462.12ms | 31.69ms | 17.4ms | 4.29ms | 4.56ms | 21.27ms | 4.07ms |
| AlexNet | 503.84ms | ms | 9.73ms | 5.66ms | 8.87ms | ms | ms |

#### Average inference time of 200 rounds
| Model   |     OpenCV.js wasm      |  OpenCV.js wasm+simd+threads | OpenCV native default | OpenCV OpenVINO | OpenCV WebNN | OpenCV.js WebNN-polyfill | OpenCV.js WebNN-Electron |
|:----------:|:-------------:|:------:|:----------:|:---------:|:---------:|:--------:|:------------:|
| GoogleNet | 862.14 ms | 51.33 ms | 10.48ms | 3.68ms | 7.44ms | 64.28ms | 24.85ms |
| SqueezeNet |  461.71 ms | 15.24 ms | 3.99ms | 1.83ms | 1.96ms | 24.74ms | 1.97ms |
| AlexNet | ms | ms | 7.56ms | 5.09ms | 7.23ms | ms | ms |

### Performance analysis
#### OpenCV native DNN module
From the performance results above, we could find that in OpenCV native DNN module (GoogleNet for example), using WebNN backend is 5ms (18.2%) faster than using default implementation. However, there is still a gap between using WebNN backend and using OpenVINO backend. I think that this is because the LRN and Dropout layers in GoogleNet is not implemented by WebNN yet, which in turn divides the graph into four sub-graphs. Then the four sub-graphs are linked with default LRN and Dropout implementations. Using such sub-graphs instead of using a whole graph to do the optimization reduces the performance of OpenCV native DNN module with WebNN backend. To the contrast, SqueezeNet's ops are all use WebNN backend except Softmax. Thus, SqueezeNet using WebNN backend is not divided into different parts and its performance is very close to the SqueezeNet using OpenVINO backend.
#### OpenCV.js DNN module
From the performance results above, we could find that Both OpenCV.js DNN module (GoogleNet as an example) using WebNN-polyfill and using WebNN-Electron are better than OpenCV.js DNN module using only wasm. They outperform the default OpenCV.js DNN module using only wasm 10.93x and 32.13x respectively. Compared with OpenCV.js DNN module with wasm+simd+threads, OpenCV.js DNN module with WebNN-polyfill is similar to it while OpenCV.js DNN module with WebNN-Electron is at least 2x faster than it. 


```
force_builders=docs,Custom
buildworker:Docs=linux-4,linux-6
build_image:Docs=docs-js:18.04
build_image:Custom=javascript
buildworker:Custom=linux-4,linux-6
```
